### PR TITLE
add sending queue for relayer

### PR DESCRIPTION
**Pull request by andrewdamelio**

### CODE REVIEW

**Review by nuevoalex — CHANGES_REQUESTED** [review]
I think this would work more clearly if we used the AsyncQueue class for our queue instead as we do in other code locations like `EthersProvider.ts` or `ContractDependenciesGnosis.ts`. That way the SIG_QUEUE could be an actual queue that recieves tasks which include the `{ tx: relayTransaction, txMetadata }` data and will queue/process them as expected without any logic needed on our part for that. In the existing implementation for example if 2 items queue up the second one will get dropped.

**Comment by andrewdamelio** [comment]
@nuevoalex with the AsyncQueue class, am i able to clear out and abort the rest of the queue, if i need to (relayer down)?

**Comment by nuevoalex** [comment]
@andrewdamelio Yup! See the `kill` method here: https://caolan.github.io/async/v3/docs.html#QueueObject

File: packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts [hunk 1]
```diff
@@ -49,6 +53,49 @@ export class ContractDependenciesGnosis extends ContractDependenciesEthers {
...
+            TransactionStatus.RELAYER_DOWN,
+          );
+        } else if (error === TransactionStatus.FAILURE) {
+          this.onTransactionStatusChanged(
+            request.txMetadata,
+            TransactionStatus.FAILURE,
+          );
+        }
+
+        await this._relayQueue.kill();
```
Comment (nuevoalex) [inline packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts]: this should only be done for the RELAYER_DOWN case I believe

**Review by nuevoalex — COMMENTED** [review]
LGTM so far!

File: packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts (hunk 1 above)
Comment (andrewdamelio) [reply packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts]: Correct -- my mistake - was just lazy testing :/

File: packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts [hunk 2]
```diff
@@ -49,6 +53,50 @@ export class ContractDependenciesGnosis extends ContractDependenciesEthers {
...
+        } else if (error === TransactionStatus.FAILURE) {
+          this.onTransactionStatusChanged(
+            request.txMetadata,
+            TransactionStatus.FAILURE,
+          );
+        }
+        throw error;
+      }
+    } else{
+      // If the Relay Service is not being used so we'll execute the TX directly
```
Comment (robert-e-davidson3) [inline packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts]:
```suggestion
      // The Relay Service is not being used so we'll execute the TX directly
```

File: packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts [hunk 3]
```diff
@@ -49,6 +53,50 @@ export class ContractDependenciesGnosis extends ContractDependenciesEthers {
...
+          );
+        } else if (error === TransactionStatus.FAILURE) {
+          this.onTransactionStatusChanged(
+            request.txMetadata,
+            TransactionStatus.FAILURE,
+          );
+        }
+        throw error;
+      }
+    } else{
```
Comment (robert-e-davidson3) [inline packages/contract-dependencies-gnosis/src/ContractDependenciesGnosis.ts]:
```suggestion
    } else {
```

**Review by nuevoalex — APPROVED** [review]
👍

**Pull request merged**